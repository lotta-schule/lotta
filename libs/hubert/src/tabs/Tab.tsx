import * as React from 'react';
import { ButtonProps } from '../button/Button';
import { NavigationButton } from '../button/NavigationButton';
import clsx from 'clsx';

import styles from './Tab.module.scss';

export type TabProps = Omit<ButtonProps, 'onClick' | 'ref'> & {
  value: string | number;
  selected?: boolean;
};

export const Tab = ({ className, children, ...props }: TabProps) => {
  return (
    <NavigationButton
      className={clsx(styles.tab, className, {
        [styles.selected]: props.selected,
      })}
      role={'tab'}
      {...props}
    >
      {children}
    </NavigationButton>
  );
};
Tab.displayName = 'Tab';
