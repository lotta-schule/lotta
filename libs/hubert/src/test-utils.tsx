import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, type RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';
import { BrowserNode, BrowserState, NodeList } from 'browser';
import { BrowserStateContext } from 'browser/BrowserStateContext';
import * as fixtures from 'test-fixtures';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement<any>,
  renderOptions: RenderOptions = {}
): ReturnType<typeof render> =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const createPromise = <T = void,>() => {
  let resolve: (value: T) => void;
  let reject: (error?: any) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve: resolve!, reject: reject! };
};

// For Testing components that use the BrowserStateContext
export type TestBrowserWrapperProps = {
  children?: React.ReactNode;
  defaultNodes?: BrowserNode[];
} & Partial<Omit<BrowserState, 'onRequestChildNodes' | 'renderNodeList'>>;
export const TestBrowserWrapper = ({
  children,
  mode = 'view-and-edit',
  defaultNodes = fixtures.browserNodes,
  moveNode = vi.fn(),
  renameNode = vi.fn(),
  deleteNode = vi.fn(),
  searchNodes = vi.fn(),
  currentPath = [],
  selected = [],
  isNodeDisabled = vi.fn(() => false),
  onSelect = vi.fn(),
  onNavigate = vi.fn(),
  setCurrentAction = vi.fn(),
  currentAction = null,
  currentSearchResults = null,
  setCurrentSearchResults = vi.fn(),
  canEdit,
  uploadClient = {
    addFile: vi.fn(),
    byState: { uploading: [], pending: [], done: [], error: [] },
    hasErrors: false,
    isSuccess: false,
    isUploading: false,
    currentProgress: null,
    currentUploads: [],
  },
}: TestBrowserWrapperProps) => {
  const [nodes, setNodes] = React.useState<BrowserNode[]>(defaultNodes);
  const [isFilePreviewVisible, setIsFilePreviewVisible] = React.useState(false);

  const onRequestChildNodes = (node: BrowserNode<'directory'> | null) =>
    nodes.filter((n) => n.parent === (node?.id ?? null));

  return (
    <BrowserStateContext.Provider
      value={{
        uploadClient,
        canEdit,
        onRequestChildNodes: async (node) =>
          onRequestChildNodes(node as BrowserNode<'directory'>),
        renderNodeList: ({ path }) => (
          <NodeList path={path} nodes={onRequestChildNodes(path.at(-1)!)} />
        ),
        currentAction,
        onSelect,
        onNavigate,
        currentSearchResults,
        currentPath,
        selected,
        mode,
        isNodeDisabled,
        setCurrentAction,
        setCurrentSearchResults,
        moveNode,
        renameNode,
        deleteNode,
        searchNodes,
        isFilePreviewVisible,
        setIsFilePreviewVisible,
        createDirectory: async (parentNode, name) => {
          setNodes((nodes) => {
            return [
              ...nodes,
              {
                id: `${nodes.length + 1}`,
                name,
                type: 'directory',
                parent: parentNode?.id ?? null,
                meta: {},
              },
            ];
          });
        },
      }}
    >
      {children}
    </BrowserStateContext.Provider>
  );
};

export * as fixtures from './test-fixtures';
