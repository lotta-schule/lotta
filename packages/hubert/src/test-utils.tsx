import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement,
  renderOptions: Omit<RenderOptions, 'wrapper'> = {}
) =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
