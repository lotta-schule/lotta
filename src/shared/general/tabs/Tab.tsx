import * as React from 'react';
import { ButtonProps } from '../button/Button';
import { NavigationButton } from '../button/NavigationButton';
import clsx from 'clsx';

import styles from './Tab.module.scss';

export type TabProps = Omit<ButtonProps, 'onClick' | 'ref'> & {
    value: string | number;
    selected?: boolean;
};

export const Tab: React.FC<TabProps> = ({ className, children, ...props }) => {
    return (
        <NavigationButton
            className={clsx(styles.tab, className)}
            role={'tab'}
            {...props}
        >
            {children}
        </NavigationButton>
    );
};
Tab.displayName = 'Tab';
