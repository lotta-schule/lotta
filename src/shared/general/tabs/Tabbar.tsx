import * as React from 'react';
import { TabProps } from './Tab';
import clsx from 'clsx';

import styles from './Tab.module.scss';

export type TabbarProps = {
    className?: string;
    value?: string | number;
    onChange?: (value: string | number) => void;
    children?: any;
};

export const Tabbar: React.FC<TabbarProps> = ({
    className,
    onChange,
    value,
    children,
}) => {
    return (
        <div className={clsx(styles.tabbar, className)} role={'tablist'}>
            {React.Children.map(
                children as any,
                (child: React.ReactElement<TabProps>, i) =>
                    child &&
                    React.cloneElement(child as any, {
                        key: i,
                        onClick: () => {
                            onChange?.(child.props.value as string);
                        },
                        selected: value === child.props.value,
                    })
            )}
        </div>
    );
};
Tabbar.displayName = 'Tabbar';
