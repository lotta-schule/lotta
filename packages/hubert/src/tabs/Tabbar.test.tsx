import * as React from 'react';
import { render } from '../test-utils';
import { Tab } from './Tab';
import { Tabbar } from './Tabbar';
import userEvent from '@testing-library/user-event';

describe('tabs/Tabbar', () => {
  it('snapshot test', () => {
    const screen = render(
      <Tabbar value={'3'} onChange={jest.fn}>
        <Tab value={'0'}>Tab1 bla bla bla</Tab>
        <Tab value={'1'}>Tab2</Tab>
        <Tab value={'2'}>Tab3</Tab>
        <Tab value={'3'}>Tab4 dingsi bumso</Tab>
        <Tab value={'4'}>Tab5</Tab>
      </Tabbar>
    );
    expect(screen.getByRole('tablist')).toMatchSnapshot();
  });

  it('should have the correct tab selected', () => {
    const screen = render(
      <Tabbar value={'2'} onChange={jest.fn}>
        <Tab value={'0'}>Tab1 bla bla bla</Tab>
        <Tab value={'1'}>Tab2</Tab>
        <Tab value={'2'}>Tab3</Tab>
        <Tab value={'3'}>Tab4 dingsi bumso</Tab>
        <Tab value={'4'}>Tab5</Tab>
      </Tabbar>
    );
    expect(screen.getByRole('tab', { selected: true })).toHaveTextContent(
      'Tab3'
    );
  });

  it('should call onChange with the correct value', async () => {
    const user = userEvent.setup();
    const onChange = jest.fn();

    const screen = render(
      <Tabbar value={'2'} onChange={onChange}>
        <Tab value={'0'}>Tab1 bla bla bla</Tab>
        <Tab value={'1'}>Tab2</Tab>
        <Tab value={'2'}>Tab3</Tab>
        <Tab value={'3'}>Tab4 dingsi bumso</Tab>
        <Tab value={'4'}>Tab5</Tab>
      </Tabbar>
    );

    await user.click(screen.getByRole('tab', { name: 'Tab5' }));
    expect(onChange).toHaveBeenCalledWith('4');
  });
});
