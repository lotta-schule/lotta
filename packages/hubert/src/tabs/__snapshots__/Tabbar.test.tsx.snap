// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs/Tabbar snapshot test 1`] = `
<div
  class="tabbar"
  role="tablist"
>
  <button
    aria-selected="false"
    class="root root root tab"
    data-variant="default"
    role="tab"
    type="button"
    value="0"
  >
    <span
      class="label label"
    >
      Tab1 bla bla bla
    </span>
  </button>
  <button
    aria-selected="false"
    class="root root root tab"
    data-variant="default"
    role="tab"
    type="button"
    value="1"
  >
    <span
      class="label label"
    >
      Tab2
    </span>
  </button>
  <button
    aria-selected="false"
    class="root root root tab"
    data-variant="default"
    role="tab"
    type="button"
    value="2"
  >
    <span
      class="label label"
    >
      Tab3
    </span>
  </button>
  <button
    aria-selected="true"
    class="root root root tab selected selected selected"
    data-variant="default"
    role="tab"
    type="button"
    value="3"
  >
    <span
      class="label label"
    >
      Tab4 dingsi bumso
    </span>
  </button>
  <button
    aria-selected="false"
    class="root root root tab"
    data-variant="default"
    role="tab"
    type="button"
    value="4"
  >
    <span
      class="label label"
    >
      Tab5
    </span>
  </button>
</div>
`;
